tests/honey_badger.rs

```diff
@@ -9,15 +10,110 @@ extern crate rand;
 
 mod network;
 
+use std::collections::BTreeMap;
 use std::iter::once;
 use std::rc::Rc;
 
 use rand::Rng;
 
-use hbbft::honey_badger::HoneyBadger;
-use hbbft::messaging::NetworkInfo;
+use hbbft::honey_badger::{self, HoneyBadger, MessageContent};
+use hbbft::messaging::{NetworkInfo, Target, TargetedMessage};
 
-use network::{Adversary, MessageScheduler, NodeUid, SilentAdversary, TestNetwork, TestNode};
+use network::{
+    Adversary, MessageScheduler, MessageWithSender, NodeUid, SilentAdversary, TestNetwork, TestNode,
+};
+
+/// An adversary whose nodes never send any messages.
+pub struct FaultyShareAdversary {
+    num_good: usize,
+    num_adv: usize,
+    adv_nodes: BTreeMap<NodeUid, Rc<NetworkInfo<NodeUid>>>,
+    scheduler: MessageScheduler,
+    share_triggers: BTreeMap<u64, bool>,
+}
+
+impl FaultyShareAdversary {
+    /// Creates a new silent adversary with the given message scheduler.
+    pub fn new(
+        num_good: usize,
+        num_adv: usize,
+        adv_nodes: BTreeMap<NodeUid, Rc<NetworkInfo<NodeUid>>>,
+        scheduler: MessageScheduler,
+    ) -> FaultyShareAdversary {
+        FaultyShareAdversary {
+            num_good,
+            num_adv,
+            scheduler,
+            share_triggers: BTreeMap::new(),
+            adv_nodes,
+        }
+    }
+}
+
+impl Adversary<HoneyBadger<usize, NodeUid>> for FaultyShareAdversary {
+    fn pick_node(
+        &self,
+        nodes: &BTreeMap<NodeUid, TestNode<HoneyBadger<usize, NodeUid>>>,
+    ) -> NodeUid {
+        self.scheduler.pick_node(nodes)
+    }
+
+    fn push_message(
+        &mut self,
+        sender_id: NodeUid,
+        msg: TargetedMessage<honey_badger::Message<NodeUid>, NodeUid>,
+    ) {
+        let NodeUid(sender_id) = sender_id;
+        if sender_id < self.num_good {
+            if let TargetedMessage {
+                target: Target::All,
+                message,
+            } = msg
+            {
+                let epoch = message.epoch();
+                // Set the trigger to simulate decryption share messages.
+                self.share_triggers.entry(epoch).or_insert(true);
+            }
+        }
+    }
+
+    fn step(&mut self) -> Vec<MessageWithSender<HoneyBadger<usize, NodeUid>>> {
+        let mut outgoing = vec![];
+        let fake_proposal = &Vec::from("X marks the spot");
+
+        for (epoch, mut trigger_set) in &mut self.share_triggers {
```

Comment:
Probably `trigger_set` doesn't need to be mutable: It's already `&mut` — no need to make it `mut &mut`.